rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,105 @@
+---
+authors: Vladimir Kochnev (marshall-lee@evilmartians.com)
+state: draft
+---
+
+# RFD 1 - Kubernetes manager
+
+## What
+
+Teleport Kubernetes Manager is a [controller for Kubernetes](https://kubernetes.io/docs/concepts/architecture/controller/) serving Teleport resources defined as [Kubernetes custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+## Why
+
+Automated bootstrap of Teleport data is not an easy task, especially when it's running in Kubernetes. One can use `tctl` from the Teleport CLI toolset to bootstrap users/roles but integrate it into something like [Helm charts](https://helm.sh/docs/topics/charts/) quickly becomes a pain. In many scenarios, it's far more convenient to have the Teleport data defined the same way as the Kubernetes resources like Deployments and Services, in a format very similar to Teleport's own format.
+
+## Details
+
+### Implementation
+
+For the task, the [Kubebuilder](https://kubebuilder.io/) framework should be used. It has great code generation capabilities and gives a complete setup to develop & deploy the manager.
+
+### Teleport Auth server discovery and authentication
+
+To write something into Teleport, the manager needs to know what server to connect to and how. There're many options on how to discover Teleport instances in the cluster but for now, let's start with the simplest one - run manager as a sidecar container in the same pod as the Teleport auth server. Containers in the pod can share the volumes, so the manager container could have access to `/var/lib/teleport/proc` database, which contains administrator credentials. This is the same way how `tctl` utility works - it doesn't require any authentication when you run it on the same machine where Teleport is running.
```

Comment:
In HA scenario with multiple auth servers running, each one will have the operator sidecar I presume? How will the operator handle HA and how will multiple operators work together? Might be helpful to explore this a bit.

Reply:
controllers written with the `kubebuilder` framework (like this one) utilise a locking mechanism. So if two operators claim to control the same resource, only the one holding the lock will receive events (afaik).

There are probably some edge-cases (for example, we don't want that second sidecar to crash-loop), but those can be sorted out I assume.

Reply:
Yes `controller-runtime` utilizes a built-in Kubernetes leader election used by all the .

Reply:
Makes sense, let's mention this in the RFD please (add a separate section on HA).